### PR TITLE
config: enforcing terminal filters are the final filter in their respective chains

### LINKED ISSUE #7767: HTTP filters that put after envoy.router don't work but envoy can still boot up with that configuration
*Title*: *HTTP filters that put after envoy.router don't work but envoy can still boot up with that configuration*

*Description*:

I'm trying to add the buffer filter to the http filters list so that we can set max request size. What I observed is that if I put the buffer filter after envoy.router filter, I don't get a 413 response  when sending a request with a large file, but envoy can still boot with that configuration. On the other hand, if I reverse the order and put envoy.router filter after the buffer filter, it works as expected.

It's weird to me that yaml configuration still compiles and envoy is able to boot up with that configuration while the router filter doesn't propagate requests to subsequent filters. I can't seem to find anything in the docs implying this behavior. At least I would expect the router filter doc be updated saying that it should be the last one in the http filters list.

*Config*:

The one that's not working:
```
...
                http_filters:
                  - name: envoy.router
                    typed_config: {}
                  - name: envoy.buffer
                    config:
                      max_request_bytes: 30000000
...
```

The one that's working:
```
...
                http_filters:
                  - name: envoy.buffer
                    config:
                      max_request_bytes: 30000000
                  - name: envoy.router
                    typed_config: {}
...
```

### PR DESCRIPTION
**Pull request by alyssawilk**
An (no longer annoyingly one-off) solution to the common problem folks run into with Envoy configs where they add their filters behind the router filter and don't get why things aren't working.  Ditto for HCM, tcp_proxy etc for L4

Risk Level: Low (except for folks with broken config)
Testing: new UT
Docs Changes: n/a
Release Notes: n/a
Fixes #7767

### CODE REVIEW

File: test/extensions/filters/network/http_connection_manager/config_test.cc [hunk 1]
```diff
@@ -166,6 +196,9 @@ stat_prefix: router
   request_headers_for_tags:
   - foo
 http_filters:
+- name: envoy.health_check
```
Comment (mattklein123) [inline test/extensions/filters/network/http_connection_manager/config_test.cc]: Why is this change needed?

Comment (alyssawilk) [reply test/extensions/filters/network/http_connection_manager/config_test.cc]: Not needed - I just wanted the comparable test to have 2 filters in the right order to verify that worked.
Can revert if you'd prefer - obviously we have enough integration tests of multiple filters

Comment (mattklein123) [reply test/extensions/filters/network/http_connection_manager/config_test.cc]: I'm fine either way. I guess I have a mild preference to revert but I don't feel strongly about it.

**Comment by alyssawilk** [comment]
Though in retrospect if we're doing tcp proxy should I also do HCM?

**Comment by alyssawilk** [comment]
I'm going to go with no, just to avoid sliding down this slippery slope any further.  I think if it causes envoy maintainer burden we can always add it later. :-)

**Comment by mattklein123** [comment]
@alyssawilk please check out my comment here https://github.com/envoyproxy/envoy/issues/7767#issuecomment-516951210 before you merge.

**Comment by zuercher** [comment]
Oh actually, should we require a terminal filter at the end of a filter chain?

**Comment by mattklein123** [comment]
> Oh actually, should we require a terminal filter at the end of a filter chain?

+1 this SGTM

**Comment by alyssawilk** [comment]
+1 on terminal filter, though that may make this a breaking change for folks.  I'll definitely add this to the version history.  I don't think it merits envoy-announce but please chime in if you think it does.

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 2]
```diff
@@ -307,7 +307,15 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
 
   const auto& filters = config.http_filters();
   for (int32_t i = 0; i < filters.size(); i++) {
-    processFilter(filters[i], i, "http", filter_factories_);
+    bool is_terminal = false;
+    processFilter(filters[i], i, "http", filter_factories_, is_terminal);
+    if (is_terminal && i != filters.size() - 1) {
```
Comment (mattklein123) [inline source/extensions/filters/network/http_connection_manager/config.cc]: This logic is repeated in multiple places (and I think not fully repeated in the upgrade path below). Can we put into a helper function potentially and use that everywhere?

**Comment by alyssawilk** [comment]
Argh.
I think to make this work I'm going to have to land this as 3 PRs - the API change without enforcement, update echo2 / envoy-filer-example, and then add enforcement.

**Comment by alyssawilk** [comment]
ok, undid the "and the last filter must be terminal" for network with some TODOs.  I'll land an update to echo2, then reinstate what I had last commit (unless there's a better way to handle the various repos)
